test/test_autograd.py

```diff
@@ -331,7 +331,8 @@ def group(l, group_size):
         # Compute the d loss / d intermediates in chunks of shard_size
         shard_size = 2
         d_intermediates = [d_i for intermediates_batch in group(intermediates, shard_size)
-                           for d_i in torch.autograd.grad(loss, intermediates_batch)]
+                           for d_i in torch.autograd.grad(loss, intermediates_batch,
+                                                          retain_graph=True)]
```

Comment:
This is because the graph has one more depth due to type cast.

Reply:
The type cast will be no-op when dtype is not passed, but still adds another node to the graph. To avoid it altogether, we'll need to provide an overload to the native function without the `dtype` kwarg.

The behavior that is surprising is that `retain_graph` not being needed for the simplest backwards cases. The documentation says the grad graph will be freed, yet this test still manages to use it more than once. This is not a problem for just this PR. This quirk will apply with any change to any derivative, when the node count goes from one to two.

Reply:
it's the addition of the None node that causes you to need retain_graph?